File: streaming-service/src/main/java/com/musicapp/streamingservice/security/JwtService.java
package com.musicapp.streamingservice.security;

import com.musicapp.streamingservice.dto.UploadDto;
import com.musicapp.streamingservice.exception.AuthException;
import io.jsonwebtoken.Claims;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import jakarta.annotation.PostConstruct;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Service;

import javax.crypto.SecretKey;

@Service
public class JwtService {
    @Value("${jwt.secret}")
    private String secretKeyString;

    private SecretKey secretKey;

    @PostConstruct
    private void constructSecretKey() {
        this.secretKey = Keys.hmacShaKeyFor(secretKeyString.getBytes());
    }

    public UploadDto validateTokenAdnGetUploadDto (String token) {
        try {
            Claims claims = Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token).getPayload();
            String id = claims.getSubject();
            String action = claims.get("action").toString();
            if (!action.equals("UPLOAD")) {
                throw new AuthException("Invalid action");
            }
            return new UploadDto(id);
        } catch (Exception e) {
            throw new AuthException("Token invalid");
        }
    }
}

File: image-service/src/main/kotlin/com/musicapps/imageservice/security/JwtService.kt
package com.musicapps.imageservice.security

import com.musicapps.imageservice.dto.UploadDto
import com.musicapps.imageservice.exception.AuthException
import io.jsonwebtoken.Jwts
import jakarta.annotation.PostConstruct
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Service
import javax.crypto.SecretKey

@Service
class JwtService(@param:Value($$"${jwt.secret}") private val secretKeyString: String) {
    private lateinit var secretKey: SecretKey

    @PostConstruct
    private fun constructSecretKey() {

    }

    fun validateTokenAndGetUploadDto(token: String): UploadDto {
        try {
            val claims = Jwts.parser().verifyWith(secretKey).build().parseSignedClaims(token).payload
            val id = claims.subject
            val action = claims["action"].toString()
            if (!action.contentEquals("UPLOAD")) {
                throw AuthException("Invalid action")
            }
            return UploadDto(id)
        } catch (e: Exception) {
            throw AuthException("Invalid token")
        }
    }
}